// @ngx-formly/core/esm5/lib/components/formly.field.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function FormlyFieldConfig() { }
if (false) {
    /**
     * The model that stores all the data, where the model[key] is the value of the field
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.model;
    /**
     * The parent field.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.parent;
    /** @type {?|undefined} */
    FormlyFieldConfig.prototype.options;
    /** @type {?|undefined} */
    FormlyFieldConfig.prototype.form;
    /**
     * The key that relates to the model. This will link the field value to the model
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.key;
    /**
     * This allows you to specify the `id` of your field. Note, the `id` is generated if not set.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.id;
    /**
     * If you wish, you can specify a specific `name` for your field. This is useful if you're posting the form to a server using techniques of yester-year.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.name;
    /**
     * This is reserved for the templates. Any template-specific options go in here. Look at your specific template implementation to know the options required for this.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.templateOptions;
    /** @type {?|undefined} */
    FormlyFieldConfig.prototype.optionsTypes;
    /**
     * An object with a few useful properties
     * - `validation.messages`: A map of message names that will be displayed when the field has errors.
     * - `validation.show`: A boolean you as the developer can set to force displaying errors whatever the state of field. This is useful when you're trying to call the user's attention to some fields for some reason.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.validation;
    /**
     * Used to set validation rules for a particular field.
     * Should be an object of key - value pairs. The value can either be an expression to evaluate or a function to run.
     * Each should return a boolean value, returning true when the field is valid. See Validation for more information.
     *
     * {
     *   validation?: (string | ValidatorFn)[];
     *   [key: string]: ((control: AbstractControl, field: FormlyFieldConfig) => boolean) | ({ expression: (control: AbstractControl, field: FormlyFieldConfig) => boolean, message: ValidationMessageOption['message'] });
     * }
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.validators;
    /**
     * Use this one for anything that needs to validate asynchronously.
     * Pretty much exactly the same as the validators api, except it must be a function that returns a promise.
     *
     * {
     *   validation?: (string | AsyncValidatorFn)[];
     *   [key: string]: ((control: AbstractControl, field: FormlyFieldConfig) => Promise<boolean>) | ({ expression: (control: AbstractControl, field: FormlyFieldConfig) => Promise<boolean>, message: string });
     * }
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.asyncValidators;
    /**
     * Can be set instead of `type` to render custom html content.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.template;
    /**
     *  It is expected to be the name of the wrappers.
     *  The formly field template will be wrapped by the first wrapper, then the second, then the third, etc.
     *  You can also specify these as part of a type (which is the recommended approach).
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.wrappers;
    /**
     * Whether to hide the field. Defaults to false. If you wish this to be conditional use `hideExpression`
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.hide;
    /**
     * Conditionally hiding Field based on values from other Fields
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.hideExpression;
    /**
     * An object where the key is a property to be set on the main field config and the value is an expression used to assign that property.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.expressionProperties;
    /**
     * This is the [FormControl](https://angular.io/api/forms/FormControl) for the field.
     * It provides you more control like running validators, calculating status, and resetting state.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.formControl;
    /**
     * You can specify your own class that will be applied to the `formly-field` component.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.className;
    /**
     * Specify your own class that will be applied to the `formly-group` component.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.fieldGroupClassName;
    /**
     * A field group is a way to group fields together, making advanced layout very simple.
     * It can also be used to group fields that are associated with the same model (useful if it's different than the model for the rest of the fields).
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.fieldGroup;
    /** @type {?|undefined} */
    FormlyFieldConfig.prototype.fieldArray;
    /**
     * This should be a formly-field type added either by you or a plugin. More information over at Creating Formly Fields.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.type;
    /**
     * Whether to focus or blur the element field. Defaults to false. If you wish this to be conditional use `expressionProperties`
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.focus;
    /**
     * An object with a few useful properties to control the model changes
     * - `debounce`: integer value which contains the debounce model update value in milliseconds. A value of 0 triggers an immediate update.
     * - `updateOn`: string event value that instructs when the control should be updated
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.modelOptions;
    /** @type {?|undefined} */
    FormlyFieldConfig.prototype.hooks;
    /**
     * @deprecated use `hooks` instead
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.lifecycle;
    /**
     * Use `defaultValue` to initialize it the model. If this is provided and the value of the model at compile-time is undefined, then the value of the model will be assigned to `defaultValue`.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.defaultValue;
    /**
     * Array of functions to execute, as a pipeline, whenever the model updates, usually via user input.
     * @type {?|undefined}
     */
    FormlyFieldConfig.prototype.parsers;
}
/**
 * @record
 */
export function ExpressionPropertyCache() { }
if (false) {
    /** @type {?} */
    ExpressionPropertyCache.prototype.expression;
    /** @type {?|undefined} */
    ExpressionPropertyCache.prototype.expressionValue;
}
/**
 * @record
 */
export function FormlyFieldConfigCache() { }
if (false) {
    /** @type {?|undefined} */
    FormlyFieldConfigCache.prototype.parent;
    /** @type {?|undefined} */
    FormlyFieldConfigCache.prototype.options;
    /** @type {?|undefined} */
    FormlyFieldConfigCache.prototype._expressionProperties;
    /** @type {?|undefined} */
    FormlyFieldConfigCache.prototype._hide;
    /** @type {?|undefined} */
    FormlyFieldConfigCache.prototype._validators;
    /** @type {?|undefined} */
    FormlyFieldConfigCache.prototype._asyncValidators;
    /** @type {?|undefined} */
    FormlyFieldConfigCache.prototype._componentRefs;
    /** @type {?|undefined} */
    FormlyFieldConfigCache.prototype._keyPath;
}
/**
 * @record
 */
export function FormlyTemplateOptions() { }
if (false) {
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.type;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.label;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.placeholder;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.disabled;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.options;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.rows;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.cols;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.description;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.hidden;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.max;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.min;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.minLength;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.maxLength;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.pattern;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.required;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.tabindex;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.readonly;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.attributes;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.step;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.focus;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.blur;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.keyup;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.keydown;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.click;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.change;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.keypress;
    /** @type {?|undefined} */
    FormlyTemplateOptions.prototype.templateManipulators;
    /* Skipping unhandled member: [additionalProperties: string]: any;*/
}
/**
 * @record
 */
export function FormlyLifeCycleFn() { }
/**
 * @record
 */
export function FormlyHookFn() { }
/**
 * @record
 * @template T
 */
export function FormlyLifeCycleOptions() { }
if (false) {
    /** @type {?|undefined} */
    FormlyLifeCycleOptions.prototype.onInit;
    /** @type {?|undefined} */
    FormlyLifeCycleOptions.prototype.onChanges;
    /** @type {?|undefined} */
    FormlyLifeCycleOptions.prototype.afterContentInit;
    /** @type {?|undefined} */
    FormlyLifeCycleOptions.prototype.afterViewInit;
    /** @type {?|undefined} */
    FormlyLifeCycleOptions.prototype.onDestroy;
    /**
     * @deprecated
     * @type {?|undefined}
     */
    FormlyLifeCycleOptions.prototype.doCheck;
    /**
     * @deprecated
     * @type {?|undefined}
     */
    FormlyLifeCycleOptions.prototype.afterContentChecked;
    /**
     * @deprecated
     * @type {?|undefined}
     */
    FormlyLifeCycleOptions.prototype.afterViewChecked;
    /* Skipping unhandled member: [additionalProperties: string]: any;*/
}
/**
 * @record
 */
export function FormlyFormOptionsCache() { }
if (false) {
    /** @type {?|undefined} */
    FormlyFormOptionsCache.prototype._checkField;
    /** @type {?|undefined} */
    FormlyFormOptionsCache.prototype._markForCheck;
    /** @type {?|undefined} */
    FormlyFormOptionsCache.prototype._buildForm;
    /** @type {?|undefined} */
    FormlyFormOptionsCache.prototype._buildField;
    /** @type {?|undefined} */
    FormlyFormOptionsCache.prototype._resolver;
    /** @type {?|undefined} */
    FormlyFormOptionsCache.prototype._injector;
    /** @type {?|undefined} */
    FormlyFormOptionsCache.prototype._hiddenFieldsForCheck;
    /** @type {?|undefined} */
    FormlyFormOptionsCache.prototype._initialModel;
}
/**
 * @record
 */
export function FormlyFormOptions() { }
if (false) {
    /** @type {?|undefined} */
    FormlyFormOptions.prototype.updateInitialValue;
    /** @type {?|undefined} */
    FormlyFormOptions.prototype.resetModel;
    /** @type {?|undefined} */
    FormlyFormOptions.prototype.formState;
    /** @type {?|undefined} */
    FormlyFormOptions.prototype.fieldChanges;
    /** @type {?|undefined} */
    FormlyFormOptions.prototype.fieldTransform;
    /** @type {?|undefined} */
    FormlyFormOptions.prototype.showError;
    /** @type {?|undefined} */
    FormlyFormOptions.prototype.parentForm;
}
/**
 * @record
 */
export function FormlyValueChangeEvent() { }
if (false) {
    /** @type {?} */
    FormlyValueChangeEvent.prototype.field;
    /** @type {?} */
    FormlyValueChangeEvent.prototype.type;
    /** @type {?} */
    FormlyValueChangeEvent.prototype.value;
    /* Skipping unhandled member: [meta: string]: any;*/
}
