// @ngx-formly/core/esm2015/lib/services/formly.builder.js
import { __decorate, __metadata, __param } from "tslib";
import { Injectable, ComponentFactoryResolver, Injector, Optional } from '@angular/core';
import { FormGroup, FormArray, FormGroupDirective } from '@angular/forms';
import { FormlyConfig } from './formly.config';
import { defineHiddenProp, reduceFormUpdateValidityCalls, observe } from '../utils';
import * as i0 from "@angular/core";
import * as i1 from "./formly.config";
import * as i2 from "@angular/forms";
let FormlyFormBuilder = class FormlyFormBuilder {
    constructor(config, resolver, injector, parentForm) {
        this.config = config;
        this.resolver = resolver;
        this.injector = injector;
        this.parentForm = parentForm;
    }
    buildForm(form, fieldGroup = [], model, options) {
        this.build({ fieldGroup, model, form, options });
    }
    build(field) {
        if (!this.config.extensions.core) {
            throw new Error('NgxFormly: missing `forRoot()` call. use `forRoot()` when registering the `FormlyModule`.');
        }
        if (!field.parent) {
            this._setOptions(field);
            reduceFormUpdateValidityCalls(field.form, () => this._build(field));
            const options = field.options;
            options.checkExpressions && options.checkExpressions(field, true);
            options.detectChanges && options.detectChanges(field);
        }
        else {
            this._build(field);
        }
    }
    _build(field) {
        if (!field) {
            return;
        }
        this.getExtensions().forEach((extension) => extension.prePopulate && extension.prePopulate(field));
        this.getExtensions().forEach((extension) => extension.onPopulate && extension.onPopulate(field));
        if (field.fieldGroup) {
            field.fieldGroup.forEach((f) => this._build(f));
        }
        this.getExtensions().forEach((extension) => extension.postPopulate && extension.postPopulate(field));
    }
    getExtensions() {
        return Object.keys(this.config.extensions).map((name) => this.config.extensions[name]);
    }
    _setOptions(field) {
        field.form = field.form || new FormGroup({});
        field.model = field.model || {};
        field.options = field.options || {};
        const options = field.options;
        if (!options._resolver) {
            defineHiddenProp(options, '_resolver', this.resolver);
        }
        if (!options._injector) {
            defineHiddenProp(options, '_injector', this.injector);
        }
        if (!options.build) {
            options._buildForm = () => {
                console.warn(`Formly: 'options._buildForm' is deprecated since v6.0, use 'options.build' instead.`);
                this.build(field);
            };
            options.build = (f) => this.build(f);
        }
        if (!options.parentForm && this.parentForm) {
            defineHiddenProp(options, 'parentForm', this.parentForm);
            observe(options, ['parentForm', 'submitted'], ({ firstChange }) => {
                if (!firstChange) {
                    options.checkExpressions(field);
                    options.detectChanges(field);
                }
            });
        }
    }
};
FormlyFormBuilder.ctorParameters = () => [
    { type: FormlyConfig },
    { type: ComponentFactoryResolver },
    { type: Injector },
    { type: FormGroupDirective, decorators: [{ type: Optional }] }
];
FormlyFormBuilder.ɵprov = i0.ɵɵdefineInjectable({ factory: function FormlyFormBuilder_Factory() { return new FormlyFormBuilder(i0.ɵɵinject(i1.FormlyConfig), i0.ɵɵinject(i0.ComponentFactoryResolver), i0.ɵɵinject(i0.INJECTOR), i0.ɵɵinject(i2.FormGroupDirective, 8)); }, token: FormlyFormBuilder, providedIn: "root" });
FormlyFormBuilder = __decorate([
    Injectable({ providedIn: 'root' }),
    __param(3, Optional()),
    __metadata("design:paramtypes", [FormlyConfig,
        ComponentFactoryResolver,
        Injector,
        FormGroupDirective])
], FormlyFormBuilder);
export { FormlyFormBuilder };
