// @clr/angular/esm2015/forms/datalist/datalist.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Input, Directive, Optional } from '@angular/core';
import { DatalistIdService } from './providers/datalist-id.service';
export class ClrDatalist {
    constructor(datalistIdService) {
        this.datalistIdService = datalistIdService;
        this.subscriptions = [];
    }
    ngAfterContentInit() {
        if (!this.datalistIdService) {
            return;
        }
        this.subscriptions.push(this.datalistIdService.idChange.subscribe(id => (this.datalistId = id)));
    }
    set id(idValue) {
        if (!!idValue && this.datalistIdService) {
            this.datalistId = idValue;
            this.datalistIdService.id = idValue;
        }
        else if (idValue) {
            this.datalistId = idValue;
        }
    }
    ngOnDestroy() {
        this.subscriptions.forEach(sub => sub.unsubscribe());
    }
}
ClrDatalist.ctorParameters = () => [
    { type: DatalistIdService, decorators: [{ type: Optional }] }
];
ClrDatalist.decorators = [
    { type: Directive, args: [{
                selector: 'datalist',
                host: {
                    '[id]': 'datalistId',
                },
            },] }
];
ClrDatalist.ctorParameters = () => [
    { type: DatalistIdService, decorators: [{ type: Optional }] }
];
ClrDatalist.propDecorators = {
    id: [{ type: Input }]
};
