// ngx-progressbar/fesm5/ngx-progressbar-http.js
import { __assign, __decorate, __param } from 'tslib';
import { InjectionToken, Optional, Inject, Injectable, NgModule } from '@angular/core';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { finalize } from 'rxjs/operators';
import { NgProgress } from 'ngx-progressbar';

var NG_PROGRESS_HTTP_CONFIG = new InjectionToken('ngProgressHttpConfig');

var NgProgressInterceptor = /** @class */ (function () {
    function NgProgressInterceptor(ngProgress, config) {
        this.ngProgress = ngProgress;
        this._inProgressCount = 0;
        this._config = {
            id: 'root',
            silentApis: []
        };
        this._config = config ? __assign(__assign({}, this._config), config) : this._config;
    }
    NgProgressInterceptor.prototype.intercept = function (req, next) {
        var _this = this;
        // Ignore by request headers
        if (req.headers.has('ignoreProgressBar')) {
            return next.handle(req.clone({ headers: req.headers.delete('ignoreProgressBar') }));
        }
        // Ignore silent api requests
        if (this.checkUrl(req)) {
            return next.handle(req);
        }
        this._inProgressCount++;
        this._progressRef = this.ngProgress.ref(this._config.id);
        if (!this._progressRef.isStarted) {
            this._progressRef.start();
        }
        return next.handle(req).pipe(finalize(function () {
            _this._inProgressCount--;
            if (_this._inProgressCount === 0) {
                _this._progressRef.complete();
            }
        }));
    };
    /**
     * Check if request is silent.
     * @param req request
     */
    NgProgressInterceptor.prototype.checkUrl = function (req) {
        var url = req.url.toLowerCase();
        var found = this._config.silentApis.find(function (u) { return url.startsWith(u); });
        return !!found;
    };
    NgProgressInterceptor.ctorParameters = function () { return [
        { type: NgProgress },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [NG_PROGRESS_HTTP_CONFIG,] }] }
    ]; };
    NgProgressInterceptor = __decorate([
        Injectable(),
        __param(1, Optional()), __param(1, Inject(NG_PROGRESS_HTTP_CONFIG))
    ], NgProgressInterceptor);
    return NgProgressInterceptor;
}());

var NgProgressHttpModule = /** @class */ (function () {
    function NgProgressHttpModule() {
    }
    NgProgressHttpModule_1 = NgProgressHttpModule;
    NgProgressHttpModule.withConfig = function (config) {
        return {
            ngModule: NgProgressHttpModule_1,
            providers: [
                { provide: NG_PROGRESS_HTTP_CONFIG, useValue: config }
            ]
        };
    };
    var NgProgressHttpModule_1;
    NgProgressHttpModule = NgProgressHttpModule_1 = __decorate([
        NgModule({
            providers: [
                { provide: HTTP_INTERCEPTORS, useClass: NgProgressInterceptor, multi: true }
            ]
        })
    ], NgProgressHttpModule);
    return NgProgressHttpModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { NG_PROGRESS_HTTP_CONFIG, NgProgressHttpModule, NgProgressInterceptor as ɵa };
