// @clr/angular/esm2015/forms/common/form.js
/**
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { ContentChildren, Directive, HostListener, Input } from '@angular/core';
import { LayoutService } from './providers/layout.service';
import { MarkControlService } from './providers/mark-control.service';
import { ClrLabel } from './label';
export class ClrForm {
    constructor(layoutService, markControlService) {
        this.layoutService = layoutService;
        this.markControlService = markControlService;
    }
    set labelSize(size) {
        this.layoutService.labelSize = size;
    }
    // Trying to avoid adding an input and keep this backwards compatible at the same time
    markAsTouched() {
        this.markControlService.markAsTouched();
    }
    onFormSubmit() {
        this.markAsTouched();
    }
}
ClrForm.ctorParameters = () => [
    { type: LayoutService },
    { type: MarkControlService }
];
ClrForm.decorators = [
    { type: Directive, args: [{
                selector: '[clrForm]',
                providers: [LayoutService, MarkControlService],
                host: {
                    '[class.clr-form]': 'true',
                    '[class.clr-form-horizontal]': 'layoutService.isHorizontal()',
                    '[class.clr-form-compact]': 'layoutService.isCompact()',
                },
            },] }
];
ClrForm.ctorParameters = () => [
    { type: LayoutService },
    { type: MarkControlService }
];
ClrForm.propDecorators = {
    labelSize: [{ type: Input, args: ['clrLabelSize',] }],
    labels: [{ type: ContentChildren, args: [ClrLabel, { descendants: true },] }],
    onFormSubmit: [{ type: HostListener, args: ['submit',] }]
};
