// @ngx-formly/material/esm5/native-select/native-select.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormlyModule } from '@ngx-formly/core';
import { FormlySelectModule } from '@ngx-formly/core/select';
import { ReactiveFormsModule } from '@angular/forms';
import { FormlyMatFormFieldModule } from '@ngx-formly/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { FormlyFieldNativeSelect } from './native-select.type';
var FormlyMatNativeSelectModule = /** @class */ (function () {
    function FormlyMatNativeSelectModule() {
    }
    FormlyMatNativeSelectModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [FormlyFieldNativeSelect],
                    imports: [
                        CommonModule,
                        ReactiveFormsModule,
                        MatInputModule,
                        FormlyMatFormFieldModule,
                        FormlySelectModule,
                        FormlyModule.forChild({
                            types: [
                                {
                                    name: 'native-select',
                                    component: FormlyFieldNativeSelect,
                                    wrappers: ['form-field'],
                                },
                            ],
                        }),
                    ],
                },] }
    ];
    return FormlyMatNativeSelectModule;
}());
export { FormlyMatNativeSelectModule };
