// @ngx-formly/core/esm5/lib/services/formly.builder.js
import { __decorate, __metadata, __param } from "tslib";
import { Injectable, ComponentFactoryResolver, Injector, Optional } from '@angular/core';
import { FormGroup, FormArray, FormGroupDirective } from '@angular/forms';
import { FormlyConfig } from './formly.config';
import { defineHiddenProp, reduceFormUpdateValidityCalls, observe } from '../utils';
import * as i0 from "@angular/core";
import * as i1 from "./formly.config";
import * as i2 from "@angular/forms";
var FormlyFormBuilder = /** @class */ (function () {
    function FormlyFormBuilder(config, resolver, injector, parentForm) {
        this.config = config;
        this.resolver = resolver;
        this.injector = injector;
        this.parentForm = parentForm;
    }
    FormlyFormBuilder.prototype.buildForm = function (form, fieldGroup, model, options) {
        if (fieldGroup === void 0) { fieldGroup = []; }
        this.build({ fieldGroup: fieldGroup, model: model, form: form, options: options });
    };
    FormlyFormBuilder.prototype.build = function (field) {
        var _this = this;
        if (!this.config.extensions.core) {
            throw new Error('NgxFormly: missing `forRoot()` call. use `forRoot()` when registering the `FormlyModule`.');
        }
        if (!field.parent) {
            this._setOptions(field);
            reduceFormUpdateValidityCalls(field.form, function () { return _this._build(field); });
            var options = field.options;
            options.checkExpressions && options.checkExpressions(field, true);
            options.detectChanges && options.detectChanges(field);
        }
        else {
            this._build(field);
        }
    };
    FormlyFormBuilder.prototype._build = function (field) {
        var _this = this;
        if (!field) {
            return;
        }
        this.getExtensions().forEach(function (extension) { return extension.prePopulate && extension.prePopulate(field); });
        this.getExtensions().forEach(function (extension) { return extension.onPopulate && extension.onPopulate(field); });
        if (field.fieldGroup) {
            field.fieldGroup.forEach(function (f) { return _this._build(f); });
        }
        this.getExtensions().forEach(function (extension) { return extension.postPopulate && extension.postPopulate(field); });
    };
    FormlyFormBuilder.prototype.getExtensions = function () {
        var _this = this;
        return Object.keys(this.config.extensions).map(function (name) { return _this.config.extensions[name]; });
    };
    FormlyFormBuilder.prototype._setOptions = function (field) {
        var _this = this;
        field.form = field.form || new FormGroup({});
        field.model = field.model || {};
        field.options = field.options || {};
        var options = field.options;
        if (!options._resolver) {
            defineHiddenProp(options, '_resolver', this.resolver);
        }
        if (!options._injector) {
            defineHiddenProp(options, '_injector', this.injector);
        }
        if (!options.build) {
            options._buildForm = function () {
                console.warn("Formly: 'options._buildForm' is deprecated since v6.0, use 'options.build' instead.");
                _this.build(field);
            };
            options.build = function (f) { return _this.build(f); };
        }
        if (!options.parentForm && this.parentForm) {
            defineHiddenProp(options, 'parentForm', this.parentForm);
            observe(options, ['parentForm', 'submitted'], function (_a) {
                var firstChange = _a.firstChange;
                if (!firstChange) {
                    options.checkExpressions(field);
                    options.detectChanges(field);
                }
            });
        }
    };
    FormlyFormBuilder.ctorParameters = function () { return [
        { type: FormlyConfig },
        { type: ComponentFactoryResolver },
        { type: Injector },
        { type: FormGroupDirective, decorators: [{ type: Optional }] }
    ]; };
    FormlyFormBuilder.ɵprov = i0.ɵɵdefineInjectable({ factory: function FormlyFormBuilder_Factory() { return new FormlyFormBuilder(i0.ɵɵinject(i1.FormlyConfig), i0.ɵɵinject(i0.ComponentFactoryResolver), i0.ɵɵinject(i0.INJECTOR), i0.ɵɵinject(i2.FormGroupDirective, 8)); }, token: FormlyFormBuilder, providedIn: "root" });
    FormlyFormBuilder = __decorate([
        Injectable({ providedIn: 'root' }),
        __param(3, Optional()),
        __metadata("design:paramtypes", [FormlyConfig,
            ComponentFactoryResolver,
            Injector,
            FormGroupDirective])
    ], FormlyFormBuilder);
    return FormlyFormBuilder;
}());
export { FormlyFormBuilder };
