// @clr/angular/esm2015/utils/popover/popover-open-close-button.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 *
 */
import { Directive, EventEmitter, HostListener, Output } from '@angular/core';
import { ClrPopoverToggleService } from './providers/popover-toggle.service';
export class ClrPopoverOpenCloseButton {
    constructor(smartOpenService) {
        this.smartOpenService = smartOpenService;
        this.subscriptions = [];
        this.openCloseChange = new EventEmitter();
        this.subscriptions.push(this.smartOpenService.openChange.subscribe(change => {
            this.openCloseChange.next(change);
        }));
    }
    handleClick(event) {
        this.smartOpenService.toggleWithEvent(event);
    }
    ngOnDestroy() {
        this.subscriptions.forEach(sub => sub.unsubscribe());
    }
}
ClrPopoverOpenCloseButton.ctorParameters = () => [
    { type: ClrPopoverToggleService }
];
ClrPopoverOpenCloseButton.decorators = [
    { type: Directive, args: [{
                selector: '[clrPopoverOpenCloseButton]',
                host: {
                    '[class.clr-smart-open-close]': 'true',
                },
            },] }
];
ClrPopoverOpenCloseButton.ctorParameters = () => [
    { type: ClrPopoverToggleService }
];
ClrPopoverOpenCloseButton.propDecorators = {
    openCloseChange: [{ type: Output, args: ['clrPopoverOpenCloseChange',] }],
    handleClick: [{ type: HostListener, args: ['click', ['$event'],] }, { type: HostListener, args: ['keydown', ['$event'],] }]
};
