// @clr/angular/esm2015/accordion/accordion.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, ContentChildren, ChangeDetectionStrategy, Input, } from '@angular/core';
import { startWith } from 'rxjs/operators';
import { AccordionService } from './providers/accordion.service';
import { ClrAccordionPanel } from './accordion-panel';
import { AccordionStrategy } from './enums/accordion-strategy.enum';
export class ClrAccordion {
    constructor(accordionService) {
        this.accordionService = accordionService;
        this.multiPanel = false;
        this.subscriptions = [];
    }
    ngOnInit() {
        this.setAccordionStrategy();
    }
    ngOnChanges(changes) {
        if (changes.multiPanel.currentValue !== changes.multiPanel.previousValue) {
            this.setAccordionStrategy();
        }
    }
    ngAfterViewInit() {
        this.subscriptions.push(this.listenForDOMChanges());
    }
    ngOnDestroy() {
        this.subscriptions.forEach(s => s.unsubscribe());
    }
    setAccordionStrategy() {
        const strategy = this.multiPanel ? AccordionStrategy.Multi : AccordionStrategy.Default;
        this.accordionService.setStrategy(strategy);
    }
    listenForDOMChanges() {
        return this.panels.changes
            .pipe(startWith(this.panels))
            .subscribe((panels) => this.accordionService.updatePanelOrder(panels.toArray().map(p => p.id)));
    }
}
ClrAccordion.ctorParameters = () => [
    { type: AccordionService }
];
ClrAccordion.decorators = [
    { type: Component, args: [{
                selector: 'clr-accordion',
                template: `<ng-content></ng-content>`,
                host: { '[class.clr-accordion]': 'true' },
                providers: [AccordionService],
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
ClrAccordion.ctorParameters = () => [
    { type: AccordionService }
];
ClrAccordion.propDecorators = {
    multiPanel: [{ type: Input, args: ['clrAccordionMultiPanel',] }],
    panels: [{ type: ContentChildren, args: [ClrAccordionPanel, { descendants: true },] }]
};
